test/replication.test.js

```diff
@@ -52,6 +52,121 @@ describe('Replication / Change APIs', function() {
     };
   });
 
+  describe('optimization check rectifyChange Vs rectifyAllChanges', function() {
+    beforeEach(function initialData(done) {
+      var data = [{name: 'foo1', lname: 'bar1'}, {name: 'foo2', lname: 'bar2'}];
+      async.auto({
+        createSourceModel: function(callback) {
+          SourceModel.create(data, callback);
+        },
+        replicate: ['createSourceModel', function(callback, results) {
+            SourceModel.replicate(TargetModel, callback);
+          }]
+      }, function(err, results) {
+        done(err);
+      });
+    });
+
+    it('rectifyOnDelete for Delete should call rectifyChange instead of rectifyAllChanges', function(done) {
+      var count = 0;
+      TargetModel.rectifyChange = function() {
+        // based on destroyAll where condition it should call rectifyChange two times
+        if (++count == 2) {
+          TargetModel.find(function(err, data) {
+            expect(data.length).to.equal(0);
+            return done(err);
+          });
+        }
+      };
+      TargetModel.rectifyAllChanges = function() {
+        return done(new Error('Should not call rectifyAllChanges'));
+      };
+
+      async.auto({
+        destroySourceModel: function(callback) {
+          SourceModel.destroyAll({'or': [{name: 'foo1'}, {lname: 'bar2'}]}, callback);
```

Comment:
Oh, I see, these two new tests are operating at a higher level, verifying that replication of Update/Delete triggers `rectifyChange` in the target model.

In that case you should not use a complex filter for `SourceModel` operation, because that filter is not passed to `TargetModel`. Always use the simplest solution that can reproduce the scenario you want to test. Otherwise readers can get confused into thinking your test is covering more than it does, as happened to me right now.

Here is what your test does under the hood currently:

```
delete all SourceModel instances matching {'or': [{name: 'foo1'}, {lname: 'bar2'}]}
  - delete instance id#1
  - delete instance id#2
replicate changes
  - update TargetModel - delete instance id#1
  - update TargetModel - delete instance id#2
```

It is good idea to keep these two high-level tests in place, please add also another one for CREATE, to get a complete coverage.

To give use higher confidence that `getIdFromWhereByModelId` works as expected for complex filters, you need to add unit-tests mocking out `SourceModel.rectifyChange` and `SourceModel.rectifyAllChanges`.

```js
var calls = [];
SourceModel.rectifyChange = function() { calls.push('rectifyChange'); };
SourceModel.rectifyAllChanges = function() { calls.push('rectifyAllChanges'); };

await SourceModel.destroyAll({'or': [{name: 'foo1'}, {lname: 'bar2'}]}

expect(calls).to.eql(['rectifyAllChanges']); // no id was specified in the query
```

Last but not least, I find the property name `lname` confusing, it looks like a typo. Could you please use `surname` or perhaps `age` instead? Also use real names instead of `bar1` and `bar2`, e.g. `John` and `Jane`.